src/transxchange/read.rs

```diff
@@ -168,15 +186,105 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
     Ok(companies)
 }
 
+fn load_commercial_physical_modes(
+    transxchange: &Element,
+) -> Result<(CommercialMode, PhysicalMode)> {
+    let raw_mode = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?
+        .try_only_child("Mode")?
+        .text();
+    let mode = MODES
+        .get(raw_mode.as_str())
+        .ok_or_else(|| {
+            format_err!(
+                "Failed to identify the commercial/physical mode from '{}'",
+                raw_mode
+            )
+        })?
+        .to_string();
+    let commercial_mode = CommercialMode {
+        id: mode.clone(),
+        name: mode.clone(),
+    };
+    let physical_mode = PhysicalMode {
+        id: mode.clone(),
+        name: mode.clone(),
+        ..Default::default()
+    };
+    Ok((commercial_mode, physical_mode))
+}
+
+fn load_line(transxchange: &Element, network_id: &str, commercial_mode_id: &str) -> Result<Line> {
+    let service = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?;
+    let id = service.try_only_child("ServiceCode")?.text();
+    let code = Some(
+        service
+            .try_only_child("Lines")?
+            .try_only_child("Line")?
+            .try_only_child("LineName")?
+            .text()
+            .trim()
+            .to_string(),
+    );
+    let name = if let Ok(description) = service.try_only_child("Description") {
+        description.text().trim().to_string()
+    } else {
+        String::from(UNDEFINED)
+    };
+    let standard_service = service.try_only_child("StandardService")?;
+    let forward_name = Some(
+        standard_service
+            .try_only_child("Destination")?
+            .text()
+            .trim()
+            .to_string(),
+    );
+    let backward_name = Some(
+        standard_service
+            .try_only_child("Origin")?
+            .text()
+            .trim()
+            .to_string(),
+    );
+    let network_id = network_id.to_string();
+    let commercial_mode_id = commercial_mode_id.to_string();
+    let line = Line {
+        id,
+        code,
+        name,
+        forward_name,
+        backward_name,
+        network_id,
+        commercial_mode_id,
+        ..Default::default()
+    };
+    Ok(line)
+}
+
 fn read_xml(transxchange: &Element, collections: &mut Collections) -> Result<()> {
     collections.datasets =
         update_validity_period_from_transxchange(&mut collections.datasets, transxchange)?;
     let network = load_network(transxchange)?;
-    if collections.networks.get(&network.id).is_none() {
-        collections.networks.push(network)?;
-    }
     let companies = load_companies(transxchange)?;
+    let (commercial_mode, physical_mode) = match load_commercial_physical_modes(transxchange) {
+        Ok(modes) => modes,
+        Err(e) => {
+            warn!("Ignore the TransXChange file: {}", e);
```

Comment:
Now useless with the application of a default mode :confused:

Reply:
Changed indeed in 9837c714c2044c279627152b80c554924f5198d8.